base/multidimensional.jl

```diff
@@ -418,6 +419,7 @@ module IteratorsMD
         @inbounds result = reshape(1:prod(dims), dims)[(I .- first.(iter.indices) .+ 1)...]
         return result
     end
+    @inline Base.getindex(iter::LinearIndices{1,R}, i::Int) where {R} = i
```

Comment:
Is it really true that this definition could apply to _all_ `LinearIndices` and not just `LinearIndices{1}`?

Reply:
It's only used for `LinearIndices` here. Do you suggest it could work for any dimensionality, i.e. that we could remove the more general method above?

Regarding switching to `IndexLinear`, I'm concerned about performance, given the comment that @timholy has put in the method above. I've tested it, and without bounds checks this test fails:
https://github.com/JuliaLang/julia/blob/c4c93ea692e9972a97a3f7149d79c92ad5d8106d/test/abstractarray.jl#L121

Given that checking bounds against the tuple of axes is going to be slow in the multidimensional case, maybe we need to keep `IndexCartesian`? I must say I don't understand where the bounds are checked currently since the method above uses `@inbounds`.

Reply:
Yeah, that method is sketchy — it's not checking bounds at all.  I'm proposing:

```diff
shell> git diff
diff --git a/base/multidimensional.jl b/base/multidimensional.jl
index 54124ed0b3..d56386b88b 100644
--- a/base/multidimensional.jl
+++ b/base/multidimensional.jl
@@ -410,16 +410,13 @@ module IteratorsMD
     LinearIndices(A::AbstractArray) = LinearIndices(CartesianIndices(A))

     # AbstractArray implementation
-    Base.IndexStyle(::Type{LinearIndices{N,R}}) where {N,R} = IndexCartesian()
+    Base.IndexStyle(::Type{LinearIndices{N,R}}) where {N,R} = IndexLinear()
     Base.axes(iter::LinearIndices{N,R}) where {N,R} = iter.indices
     Base.size(iter::LinearIndices{N,R}) where {N,R} = length.(iter.indices)
-    @inline function Base.getindex(iter::LinearIndices{N,R}, I::Vararg{Int, N}) where {N,R}
-        dims = length.(iter.indices)
-        #without the inbounds, this is slower than Base._sub2ind(iter.indices, I...)
-        @inbounds result = reshape(1:prod(dims), dims)[(I .- first.(iter.indices) .+ 1)...]
-        return result
+    @inline function Base.getindex(iter::LinearIndices, i::Int)
+        @boundscheck checkbounds(iter, i)
+        i
     end
-    @inline Base.getindex(iter::LinearIndices{1,R}, i::Int) where {R} = i
 end  # IteratorsMD
```

Amazingly, it seems to have _identical performance_ comparing `srand(0); @btime L[i,j,k] setup=(i=rand(1:3);j=rand(1:4);k=rand(1:5))` with `L = LinearIndices(rand(3,4,5))`.  And it fixes this case:

```
# This PR + my patch above:
julia> L = LinearIndices(rand(3,4,5));
julia> srand(0); @btime L[i,j,k] setup=(i=rand(1:3);j=rand(1:4);k=rand(1:5))
  31.524 ns (0 allocations: 0 bytes)
22
julia> L[3,4,6]
ERROR: BoundsError: attempt to access 3×4×5 LinearIndices{3,Tuple{Base.OneTo{Int64},Base.OneTo{Int64},Base.OneTo{Int64}}} at index [3, 4, 6]

# A 7-day old master: bf45de8c33 (2018-03-26 16:43 UTC)
julia> srand(0); @btime L[i,j,k] setup=(i=rand(1:3);j=rand(1:4);k=rand(1:5))
  31.048 ns (0 allocations: 0 bytes)
22
julia> L[3,4,6]
72
```

Reply:
Yeah, that's what I had in mind too. Glad to hear it doesn't affect performance. I've updated the PR.